SimCalorimetry/HGCalAssociatorProducers/plugins/SimClusterAssociatorByEnergyScoreProducer.cc

```diff
@@ -0,0 +1,71 @@
+
+// user include files
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/global/EDProducer.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+#include "FWCore/Utilities/interface/ESGetToken.h"
+
+#include "SimDataFormats/Associations/interface/LayerClusterToSimClusterAssociator.h"
+#include "SimClusterAssociatorByEnergyScoreImpl.h"
+
+class SimClusterAssociatorByEnergyScoreProducer : public edm::global::EDProducer<> {
+public:
+  explicit SimClusterAssociatorByEnergyScoreProducer(const edm::ParameterSet &);
+  ~SimClusterAssociatorByEnergyScoreProducer() override;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions &descriptions);
+
+private:
+  void produce(edm::StreamID, edm::Event &, const edm::EventSetup &) const override;
+  edm::EDGetTokenT<std::vector<float>> clustersMask_;
+  edm::EDGetTokenT<std::unordered_map<DetId, const HGCRecHit *>> hitMap_;
+  edm::ESGetToken<CaloGeometry, CaloGeometryRecord> caloGeometry_;
+  const bool hardScatterOnly_;
+  std::shared_ptr<hgcal::RecHitTools> rhtools_;
+};
+
+SimClusterAssociatorByEnergyScoreProducer::SimClusterAssociatorByEnergyScoreProducer(const edm::ParameterSet &ps)
+    : clustersMask_(consumes<std::vector<float>>(ps.getParameter<edm::InputTag>("LayerClustersInputMask"))),
+      hitMap_(consumes<std::unordered_map<DetId, const HGCRecHit *>>(ps.getParameter<edm::InputTag>("hitMapTag"))),
+      caloGeometry_(esConsumes<CaloGeometry, CaloGeometryRecord>()),
+      hardScatterOnly_(ps.getParameter<bool>("hardScatterOnly")) {
+  rhtools_.reset(new hgcal::RecHitTools());
+
+  // Register the product
+  produces<hgcal::LayerClusterToSimClusterAssociator>();
+}
+
+SimClusterAssociatorByEnergyScoreProducer::~SimClusterAssociatorByEnergyScoreProducer() {}
+
+void SimClusterAssociatorByEnergyScoreProducer::produce(edm::StreamID,
+                                                        edm::Event &iEvent,
+                                                        const edm::EventSetup &es) const {
+  edm::ESHandle<CaloGeometry> geom = es.getHandle(caloGeometry_);
+  rhtools_->setGeometry(*geom);
+
+  const auto &inputClusterMask = iEvent.get(clustersMask_);
+  const std::unordered_map<DetId, const HGCRecHit *> *hitMap = &iEvent.get(hitMap_);
+
+  auto impl = std::make_unique<SimClusterAssociatorByEnergyScoreImpl>(
+      iEvent.productGetter(), hardScatterOnly_, inputClusterMask, rhtools_, hitMap);
+  auto toPut = std::make_unique<hgcal::LayerClusterToSimClusterAssociator>(std::move(impl));
+  iEvent.put(std::move(toPut));
+}
+
+void SimClusterAssociatorByEnergyScoreProducer::fillDescriptions(edm::ConfigurationDescriptions &cfg) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("LayerClustersInputMask", edm::InputTag("ticlTrackstersTrkEM"));
```

Comment:
Any particular reason why you picked up that specific mask: `ticlTrackstersTrkEM`?

Reply:
More in general: would it make sense to have a unique, cumulative association instead of creating several associations based on the `LayerCluster` masks?
Those associations are time and memory consuming, and, in most cases, they will end up with quite a number of overlapping entries.

Reply:
@rovere I just picked one in the beginning since I intended to use them all. I was thinking that in these times that you make a lot of development in TICL it will be helpful to have information for all iterations. If this is not the case then we can drop that and have a  unique association.

Reply:
@apsallid  thanks for the prompt feedback. We could have information on all iterations, separately, even with a unique associator between the full set of `LayerClusters` and `SimClusters`, w/o creating as many as the iterations we will end up building.